### PR TITLE
Add support for parsing direct conversations

### PR DESCRIPTION
**Pull request by cerber-os**
### What's changed
- Added support for 1-1 private conversations (`-u` argument)

### Testing method
Created backup of random private conversation and verified its content with the data displayed on Wire Desktop app.

### Resolved
PR implements feature suggested in #2

### CODE REVIEW

File: wirebackupparser/groups.py [hunk 1]
```diff
@@ -31,19 +32,30 @@ def getMembers(self):
 
 
 class Groups:
+    """
+    Despite what name might suggest, this class handles both direct messages and group conversations.
+    Don't blame me, this is how Wire developers decided to implement backup format
+    """
```
Comment (Lachcim) [inline wirebackupparser/groups.py:38]: Wire refers to these entities as "conversations". There are 4 types of conversations: `GROUP`, `SELF`, `ONE2ONE` and `CONNECT`. I propose renaming every instance of "group" to "conversation", except where relevant.

File: wirebackupparser/groups.py [hunk 2]
```diff
@@ -31,19 +32,30 @@ def getMembers(self):
...
-                                     backupOwner=backup.backupOwner))
+            if conv.get('type') == ENT_TYPE_GROUP:
+                self.groups.append(Group(conv.get('id'),
+                                         conv.get('name'),
+                                         conv.get('creator'),
+                                         conv.get('others'),
+                                         backupOwner=backup.backupOwner))
+            elif conv.get('type') == ENT_TYPE_DIRECT:
+                self.directGroups.append(Group(conv.get('id'),
+                                               name="Direct conversation",
```
Comment (Lachcim) [inline wirebackupparser/groups.py:52]: The conversation name is relevant to the report and should not be neglected. A proper name can be obtained via `getUserByID(...).name`. Note: do not use `conv.get('name')`, those are not unique and depend on who started the conversation (i.e. sent a connection request).

File: wirebackupparser/groups.py [hunk 3]
```diff
@@ -31,19 +32,30 @@ def getMembers(self):
...
+                                               creator=None,
```
Comment (Lachcim) [inline wirebackupparser/groups.py:53]: This information is not meaningless, see above

File: wirebackupparser/groups.py [hunk 4]
```diff
@@ -53,3 +65,12 @@ def getGroupByName(self, name):
 
     def getGroups(self):
         return self.groups
+
+    def getDirectGroupByUserID(self, userID):
+        for directGroup in self.directGroups:
+            if directGroup.members[0] == userID:
+                return directGroup
+        raise KeyError("Unknown direct group with member {}".format(userID))
```
Comment (Lachcim) [inline wirebackupparser/groups.py:73]: I see this as an attempt to gracefully fail when a user lookup fails, but currently this only works for resolved usernames with whom a connection hasn't been established (and thus there isn't a one-to-one conversation).

**Review by Lachcim — CHANGES_REQUESTED** [review]
Minor issues with error handling and conversation metadata parsing. Additional code refactoring (rename "group" to "conversation") suggested.

**Comment by Lachcim** [comment]
Two remarks regarding the current state of code:

* We are throwing exceptions in response to erroneous user input, but we are not catching them anywhere. A top-level try-except statement can be added to gracefully display error messages.
* We inform the user of ambiguous usernames, but we provide no method of resolving this ambiguity. I propose implementing additional command-line arguments (`-gid`, `-uid`?) for specifying group and user IDs explicitly. Also handles text encoding issues.

File: wirebackupparser/events.py [hunk 5]
```diff
@@ -23,6 +23,13 @@ def getUserByID(uid):
     raise KeyError("Cannot find user {} in provided dict".format(uid))
 
 
+def getUserByName(name):
+    result = [user for user in Users if user.name == name]
+    if len(result) > 1:
+        raise KeyError("Ambigious username {} - found {} times".format(name, len(result)))
+    return result[0]
```
[the remark this answers is not in the document]
Comment (cerber-os) [reply wirebackupparser/events.py:32]: Fixed by 288dac4

File: wirebackupparser/groups.py (hunk 1 above)
Comment (cerber-os) [reply wirebackupparser/groups.py]: Good idea - commit 288dac4 renamed classes `Group(s)` to `Conversation(s)`

File: wirebackupparser/groups.py (hunk 2 above)
Comment (cerber-os) [reply wirebackupparser/groups.py]: Implemented in 288dac4

File: wirebackupparser/groups.py (hunk 3 above)
Comment (cerber-os) [reply wirebackupparser/groups.py]: The same as above

File: wirebackupparser/groups.py (hunk 4 above)
Comment (cerber-os) [reply wirebackupparser/groups.py]: This exception is raised when user with provided nickname is a member of one of the group but not direct conversation.
In such cases, his nickname is properly translated to `userID` (\`cause he exists in `events.json`) but no direct conversation is present in `self.directGroup` list.

Comment (Lachcim) [reply wirebackupparser/groups.py]: Correct, that's what I meant

**Comment by cerber-os** [comment]
> We are throwing exceptions in response to erroneous user input, but we are not catching them anywhere. A top-level try-except statement can be added to gracefully display error messages.

Good idea, but such feature is out of the scope of this pull request - consider creating a new issue (and possibly assign yourself).

> We inform the user of ambiguous usernames, but we provide no method of resolving this ambiguity. I propose implementing additional command-line arguments (`-gid`, `-uid`?) for specifying group and user IDs explicitly.

As far as I know, Wire prohibits ambiguous usernames by showing `Nickname is already taken` message on settings page. Therefore, I'm not sure if such options would be in any way useful in real applications.

The checks in Python code are a bit excessive, but should provide some leads when dealing with damaged/improper backup file.

> Also handles text encoding issues.

Could you elaborate?

**Comment by Lachcim** [comment]
I was able to trigger a `KeyError: 'Ambigious username 3na10 - found 2 times'` by renaming an account in the preferences screen. If you believed this to be impossible, why check against it with an if statement?
Annotation 2020-08-03 205109

Encoding issues appear when you try to type non-ASCII characters in various terminal environments. For example, on Windows, the command `echo zażółć gęślą jaźń > test.txt` outputs `zaz¢lc gesla jazn` when the codepage is set to 437. This may prohibit some users from querying certain user and group names.

**Comment by cerber-os** [comment]
Oh, you're right - Wire disallows duplicated nicknames (i.e. those prefixed with `@`), but seems to work fine with not ambiguous usernames (thanks, Wire dev team!). I'll try to add some kind of selection when more than one matching user was found.

As for the encoding issues: if your terminal doesn't support Unicode after almost 30 years since the first release of ISO10646 standard, I kindly suggest to change it...

**Comment by cerber-os** [comment]
Now, when non-unique username is provided the following message will be displayed:
```txt
./parser.py -f wire.backup -u Smith
Loading backup...
Parsing backup...
Available IDs for username Smith:
	- 4c676a26-e5e9-44ae-a738-61694cc693d4
	- 43ae3b0a-6549-47b3-9e9d-43978f54fa9d
'Ambiguous username Smith - found 2 times'
[+] Try again with option -uid
```

Newly added option `--uid` allows to select one of the available IDs:
```sh
./parser.py -f wire.backup --uid 43ae3b0a-6549-47b3-9e9d-43978f54fa9d
```

**Review by Lachcim — APPROVED** [review]
Excellent work as expected of Mr eng. Wieczorek himself. Thank you so much for this exceptional contribution. You are a scholar and a gentleman.

**Comment by h0adp0re** [comment]
Fantastic, thank you!

**Pull request merged**